Added recipe to have an interactive shell when using docker

boss6825 opened the pull request:
Fixes #1960
Supersedes and closes #1967


## Issue number - 1960




<!-- readthedocs-preview plone6 start -->
----
📚 Documentation preview 📚: https://plone6--2004.org.readthedocs.build/

<!-- readthedocs-preview plone6 end -->

boss6825 committed:
added docker recipe for interactive shell

boss6825 committed:
Clean up blank lines in index.md

Removed unnecessary blank lines from the index.md file.

erral committed:
Merge branch '6.0' into 6.0

stevepiercy commented:
> Above images are preview of documentation I added
> 
> 📚 Documentation preview 📚: https://plone6--2004.org.readthedocs.build/

FYI, we have automatic pull request previews provided by Read the Docs, so in the future, after RTD completes its build, the link will work and you don't need to add a screenshot of your local build.

You can also add a link in your description to the affected pages with anchors to specific sections.

https://plone6--2004.org.readthedocs.build/install/containers/recipes/index.html#open-an-interactive-debugging-shell-when-docker-compose

stevepiercy commented on docs/install/containers/recipes/index.md:
```diff
@@ -170,3 +170,80 @@ docker compose run backend pack
 The above command assumes that the service that runs the Plone instance is named `backend`.
 Otherwise replace `backend` with your container's name.
 
+## Open an interactive debugging shell when Docker Compose
```
```suggestion
## Open an interactive debugging shell with Docker Compose
```

stevepiercy commented on docs/install/containers/recipes/index.md:
```diff
@@ -170,3 +170,80 @@ docker compose run backend pack
 
+## Open an interactive debugging shell when Docker Compose
+
+When you need to debug a running Plone site, you must run a second Zope instance connected to the same database. This is required because an interactive console starts its own WSGI process. 
```
Per https://6.docs.plone.org/contributing/documentation/authors.html#plone-documentation-styleguide, one sentence per line.

```suggestion
When you need to debug a running Plone site, you must run a second Zope instance connected to the same database.
This is required because an interactive console starts its own WSGI process. 
```

stevepiercy commented on docs/install/containers/recipes/index.md:
```diff
@@ -170,3 +170,80 @@ docker compose run backend pack
+This recipe uses two Zope services sharing a single ZEO instance.  
+It allows you to open an interactive shell without stopping the running site.
+
+### Example docker-compose.yml
```
```suggestion
### Example `docker-compose.yml`
```

stevepiercy commented on docs/install/containers/recipes/index.md:
```diff
@@ -170,3 +170,80 @@ docker compose run backend pack
+
+### Example docker-compose.yml
+
+Save the following configuration as `docker-compose.yml`.
```
```suggestion
Save the following configuration as {file}`docker-compose.yml`.
```

stevepiercy commented on docs/install/containers/recipes/index.md:
```diff
@@ -170,3 +170,80 @@ docker compose run backend pack
+
+volumes:
+  data: {}
+
```
```suggestion
```

stevepiercy commented on docs/install/containers/recipes/index.md:
````diff
@@ -170,3 +170,80 @@ docker compose run backend pack
+
+```
+
+Now start the docker in detached mode 
````
```suggestion
Now start the docker in detached mode.
```

stevepiercy commented on docs/install/containers/recipes/index.md:
````diff
@@ -170,3 +170,80 @@ docker compose run backend pack
+
+Open the interactive shell
+
+```bash
````
Use generic shell for the lexer.

Also I like the line breaks you use, which make the commands more readable.

````suggestion
```shell
docker compose up -d
```

If the site does not yet exist in the data volume, create it.
It will then be available for future sessions.

```shell
docker exec \
  -e SITE_ID="Plone" \
  zope ./docker-entrypoint.sh create-classic
```

Open the interactive shell.

```shell
````

stevepiercy commented on docs/install/containers/recipes/index.md:
```diff
@@ -170,3 +170,80 @@ docker compose run backend pack
+This command starts a separate Zope instance connected to the same database.
+
+You cannot debug a running FileStorage-based setup because the database file is locked.
+Use ZEO or RelStorage to open an interactive shell while the site is running.
```
```suggestion
Instead, use ZEO or RelStorage to open an interactive shell while the site is running.
```

stevepiercy reviewed (changes requested):
This is really good. Just a few minor grammar and style fixes, and a little MyST markup improvements. Then a final technical review, and it can be merged. Thank you!

erral reviewed (commented):
Thanks!

There is no need to have another service configured in the `docker-compose.yml` file.

It is enough to run the following command to get the interactive command:

```
docker compose run --rm backend console
```

And it will drop into a debug console where the `app` variable holds the application instance:

```
user@server.com:~/docker/deploy(main)$ docker compose run --rm backend console
[+] Creating 1/0
 ✔ Container zeo  Running                                                                                                                                               0.0s 
/app/lib/python3.12/site-packages/plone/app/theming/__init__.py:3: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.
  __import__("pkg_resources").declare_namespace(__name__)
Starting debugger (the name "app" is bound to the top-level Zope object)
>>> app
<Application at >
>>> app.Plone
<PloneSite at /Plone>
```

boss6825 committed:
Update docs/install/containers/recipes/index.md

Co-authored-by: Steve Piercy <web@stevepiercy.com>

boss6825 committed:
Update docs/install/containers/recipes/index.md

Co-authored-by: Steve Piercy <web@stevepiercy.com>

boss6825 committed:
Update docs/install/containers/recipes/index.md

Co-authored-by: Steve Piercy <web@stevepiercy.com>

boss6825 committed:
Update docs/install/containers/recipes/index.md

Co-authored-by: Steve Piercy <web@stevepiercy.com>

boss6825 committed:
Update docs/install/containers/recipes/index.md

Co-authored-by: Steve Piercy <web@stevepiercy.com>

boss6825 committed:
Update docs/install/containers/recipes/index.md

Co-authored-by: Steve Piercy <web@stevepiercy.com>

boss6825 committed:
Update docs/install/containers/recipes/index.md

Co-authored-by: Steve Piercy <web@stevepiercy.com>

boss6825 committed:
Update docs/install/containers/recipes/index.md

Co-authored-by: Steve Piercy <web@stevepiercy.com>

stevepiercy commented:
@boss6825 would you please take a look at @erral's suggestion, try it out to verify it works for you, and revise the PR accordingly? He knows a lot more about this topic than I do.

Also, future tip, you can use the "Add suggestion to batch" button when updating your PR on GitHub to reduce the number of notifications. See https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes.

If you don't see that button, you might need to switch to the "old" GitHub version. See https://github.com/orgs/community/discussions/175664.

boss6825 commented:
@stevepiercy i am modifying current version based on @erral 's feedback, 
also for future PR's , I will use  "Add to batch ". 
Sorry for those annoying multiple notifications you got 🥲.

boss6825 committed:
Revised recipe to have an interactive shell when using docker

Updated the section on opening an interactive debugging shell with Docker Compose. Removed the example docker-compose.yml configuration and streamlined the instructions for running a debugging console.

stevepiercy reviewed (approved):
This passes my editorial review, but it needs a technical review. @erral @davisagli this should be quick. Thank you!

stevepiercy committed:
Merge branch '6.0' into 6.0
